User can block pw from another user

Mariusz08 opened the pull request:
Od teraz użytkownik może zablokować wiadomości prywatne od innego użytkownika. Do tego celu służy specjalny button w profilu. Blokada działa w dwie strony.

awaluk commented:
Nie działa mi to wcale. Przy próbie wejścia na dowolny profil użytkownika dostaje błąd:
> A Question2Answer database query failed when generating this page. A full description of the failure is available in the web server's error log file.

Nie szukałem już żadnego loga, ale widzę że są tam odwołania do jakiejś nowej tabeli, a nie widzę aby gdziekolwiek była tworzona.

Mariusz08 commented:
Chyba mam jakąś sklerozę. Specjalnie zapisałem na kartce, żeby napisać o nowej tabeli a zapomniałem... Do działania skryptu potrzebna jest nowa, osobna tabela (w niej zapisuję rekordy odnośnie blokowawnia wiadomości). Tabela jest generowana przy ponownym wywołaniu podstrony **/install**, tam się przechodzi przez cały proces a po przejściu będzie nowa tabela. Niemniej jednak, jeśli wywołamy ten skrypt cała baza danych zostanie utracona, a więc na chwilę obecną radziłbym po prostu ręcznie utworzyć taką tabelę w MySQL. Zaraz zaktualizuję dump bazy danych.

awaluk commented:
A to tworzenie na `/install` jest gdzieś zadeklarowane w kodzie? Bo jak pisałem, nie widzę aby nigdzie ta tabela była tworzona.

awaluk commented:
Ok, teraz działa ;)

- Czemu można zablokować PW od samego siebie? To raczej bez sensu.
- Problem tego typu: użytkownik blokuje wiadomości prywatne od kogoś z administracji. Ten jednak korzystając z podwyższonych uprawnień nadal może do niego pisać i to jest dobre. Tyle ze ta osoba nie może tego wyświetlić, dostaje powiadomienie o wiadomości i gdy klika żeby zobaczyć jej szczegóły to dostaje błąd że strony nie ma. Nie wiem jak to dobrze i łatwo rozegrać. Idealnie byłoby chyba jakby w przypadku blokady w skrzynce odbiorczej wyświetlane i odblokowane były tylko wiadomości od administracji, podobnie widok pojedynczej konwersacji też dozwolony tylko dla osób z administracji.
- Myślę że przydała by się lista wszystkich zablokowanych (gdzieś na profilu, tak jak np. lista ulubionych), ułatwiłoby to sprawdzenie kogo zablokowaliśmy i zapomnieliśmy np.

Mariusz08 commented:
> Czemu można zablokować PW od samego siebie? To raczej bez sensu.

Ups.... Zapomniałem o tym...

> Problem tego typu: użytkownik blokuje wiadomości prywatne od kogoś z administracji. Ten jednak korzystając z podwyższonych uprawnień nadal może do niego pisać i to jest dobre. Tyle ze ta osoba nie może tego wyświetlić, dostaje powiadomienie o wiadomości i gdy klika żeby zobaczyć jej szczegóły to dostaje błąd że strony nie ma. Nie wiem jak to dobrze i łatwo rozegrać. Idealnie byłoby chyba jakby w przypadku blokady w skrzynce odbiorczej wyświetlane i odblokowane były tylko wiadomości od administracji, podobnie widok pojedynczej konwersacji też dozwolony tylko dla osób z administracji.

Chyba można po prostu do ifa dodać warunek o rodzaju uprawnień (nawet jeśli administracja to możesz wyświetlić daną konwersację), poprawię to jutro

> Myślę że przydała by się lista wszystkich zablokowanych (gdzieś na profilu, tak jak np. lista ulubionych), ułatwiłoby to sprawdzenie kogo zablokowaliśmy i zapomnieliśmy np.

Myślałem o tym, ale zrealizowanie tego jest dosyć trudne, ponieważ:
a). nie wiem jak stworzyć stronę bezpośrednio korzystając z rdzenia Q2A (na pewno problem z routingiem [masa funkcji które przekierowują do innej itd.])
b). można stworzyć odpowiednią podstronę korzystając z pluginu aczkolwiek IMO trochę bez sensu - same operacje na buttonach czy  bazie mamy w kodzie źródłowym, a samą stronę mamy  z pluginu - coś tutaj nie pasuje.

awaluk commented:
Co do ostatniego no to tak, plugin wydaje się bezsensowny. Nie wiem, trzeba by popróbować/pomyśleć. W sensie użytkowym taka funkcja wydaje się bardzo potrzebna bo można od razu sprawdzić kogo się kiedyś tam zablokowało.

Mariusz08 commented:
Poprawione

event15 commented on dump/forum.sql:
```diff
@@ -1741,6 +1741,18 @@ INSERT INTO `qa_words` (`wordid`, `word`, `titlecount`, `contentcount`, `tagword
+-- Struktura tabeli dla tabeli `qa_blockedPw`
+--
+
+CREATE TABLE IF NOT EXISTS `qa_blockedPw` (
```
w sql nie nazywamy tabel mieszanym camelCase z innymi stylami. Przy okazji nie powinno się stosować notacji camelCase w SQL

event15 commented on forum/qa-include/pages/message.php:
```diff
@@ -1,257 +1,258 @@
-				$messagesent = true;
+    // check for messages or errors
+    $state = qa_get_state();
+    $messagesent = $state == 'message-sent';
```
powinno się stosować a) porównywanie strict (===) b) yoda condition

Comandeer replied:
> b) yoda condition

Czy mógłbyś podać jakiś standard lub zalecenia, które faktycznie zalecają taki styl pisania warunków? Osobiście udało mi się znaleźć jedynie forka PSR (ach, te standardy, zawsze podwójne…), który wręcz odradza używania → https://github.com/php-fig-rectified/fig-rectified-standards/blob/master/PSR-2-R-coding-style-guide-additions.md#comparison

awaluk replied:
Tak z ciekawości: dlaczego "powinno" się stosować yoda condition? Jakiś standard to określa? Bo nie wydaje mi się abyśmy coś takiego przyjęli i później wychodzą niejasności bo każdy robi inaczej.

event15 replied:
@Comandeer nie jest to standard określony w, o ile wiem, żadnym PSR. Jednakże w każdej firmie w której byłem był to standard wewnętrzny. Jasne, można powiedzieć, że to jest wydziwianie. Jednak z czasem uważam używanie Yoda conditions za coś ważnego, coś co jest bardziej działaniem prewencyjnym niż jakościowym.
W pewien sposób ta problematyka jest tu omówiona: https://knowthecode.io/yoda-conditions-yoda-not-yoda

Odnośnie "additions" to jak ze wszystkim, z "human readable" mógłbym dyskutować - najpewniej nie tylko ja. To problem pokroju taby czy spacje. 
Na podstawie samego niezachowania yoda conditions nie odrzuciłbym MR - to tylko komentarz, a MR został odrzucony przez inne rzeczy, takie jak np stosowanie gwiazdek w sqlkach.

Comandeer replied:
> Jednakże w każdej firmie w której byłem był to standard wewnętrzny.

A u mnie zawsze jako standard wewnętrzny jest zakaz yoda conditions ;)

event15 replied:
Dlatego mówię - to temat pokroju "taby czy spacje". Sugeruję to, co uznaję za dobre. Myślę, że to jest mało istotne. Nie ma sensu wszczynać większych dyskusji - po prostu mechanicznie zauważam brak yoda conditions w trakcie code review, o czym informuję - tak robię w pracy, tak też robię tu ;)

event15 commented on forum/qa-include/pages/message.php:
```diff
@@ -1,257 +1,258 @@
+    // check for messages or errors
+    $state = qa_get_state();
+    $messagesent = $state == 'message-sent';
+    if ($state == 'email-error')
```
powinno się stosować a) porównywanie strict (===) b) yoda condition

event15 commented on forum/qa-include/pages/message.php:
```diff
@@ -1,257 +1,258 @@
+                require_once QA_INCLUDE_DIR.'db/messages.php';
+                require_once QA_INCLUDE_DIR.'app/emails.php';
+
+                if (qa_opt('show_message_history'))
```
niezachowanie standardów PSR = po każdym bloku if musi być otworzona klamra

event15 commented on forum/qa-include/pages/user-profile.php:
```diff
@@ -921,7 +925,10 @@
         $option = qa_opt('allow_private_messages');
-        if ( $option && isset($loginuserid) && ($loginuserid != $userid) && !($useraccount['flags'] & QA_USER_FLAGS_NO_MESSAGES) && !$userediting ) {
+        
+        $isBlocked = qa_db_query_sub('SELECT * FROM ^blockedPw WHERE fromUserId = # AND toUserId = # OR fromUserId = # AND fromUserId = #', $loginuserid, $useraccount['userid'], $useraccount['userid'], $loginuserid);
```
nie jestem pewien, czy potrzeba nam robić select *, czy nie wystarczy tylko wyselekcjonować odpowiednie kolumny

event15 commented on forum/qa-include/pages/user-profile.php:
```diff
@@ -921,7 +925,10 @@
+        
+        $isBlocked = qa_db_query_sub('SELECT * FROM ^blockedPw WHERE fromUserId = # AND toUserId = # OR fromUserId = # AND fromUserId = #', $loginuserid, $useraccount['userid'], $useraccount['userid'], $loginuserid);
+
+        if ($isBlocked->num_rows === 0 && $option && isset($loginuserid) && ($loginuserid != $userid) && !($useraccount['flags'] & QA_USER_FLAGS_NO_MESSAGES) && !$userediting ) {
```
tu powinno być yoda condition, poza tym bardzo skomplikowany ten warunek -> warto go do osobnej metody a najlepiej to w ogóle porozbijać na kilka metod chociazby same warunki

Mariusz08 replied:
a) Może trochę się czepiam ale nie można mówić o metodach - kod nie jest obiektowy. Co najwyżej funkcje
b) w jaki sposób mogę rozbić ten warunek? Raczej tworzenie X funkcji jest bez sensu. Mam pomysł, aby po prostu do zmiennej przypisać warunek w taki sposób
```php
$isBlockedBool = $isBlocked->num_rows === 0;
``` 
a potem w ifie
```php
if ($isBlockedBool && ...) {}
```
Mogę tak zrobić?

event15 replied:
Metoda i funkcja to jest to samo, są to synonimy, a podział został wprowadzony tylko i wyłącznie dla twardych głów ze szkół średnich i wyższych. Semtantycznie są to te same twory.

Mariusz08 replied:
Hmm, no ok nie wiedziałem gdziekolwiek nie czytałem była omówiona różnica między tymi tworami. Niemniej jednak zręcznie udało Ci się ominąć udzielenia odpowiedzi na pytanie - czy taki sposób skrócenia ifa jest ok ;).

event15 commented on forum/qa-include/pages/user-profile.php:
```diff
@@ -2026,7 +2029,18 @@
 
 
-
+    $canBlock = qa_db_query_sub('SELECT * FROM ^blockedPw WHERE fromUserId = # AND toUserId = #', $loginuserid, $useraccount['userid']);
```
kolejny raz - nie chcę używania gwiazdek

event15 commented on forum/qa-include/pages/user-profile.php:
```diff
@@ -2026,7 +2029,18 @@
 
-
+    $canBlock = qa_db_query_sub('SELECT * FROM ^blockedPw WHERE fromUserId = # AND toUserId = #', $loginuserid, $useraccount['userid']);
+    if ($canBlock->num_rows == 0 && qa_get_logged_in_userid() != null && $loginuserid !== $useraccount['userid']) {
```
yoda condition oraz podział tych warunków na mniejsze

Mariusz08 commented:
Poprawiłem (mój kod)

awaluk commented:
Będąc userem blokuje prywatne wiadomości od admina. Jako userowi znika mi guzik wysyłania PW do admina, spoko. Tyle że mogę ręcznie wejść na `/message/admin` i bez problemu wiadomość wysłać. Z kolei gdy to admin zablokuje usera, user może nadal prywatne wiadomości do niego wysyłać, znika mu jedynie guzik. Patrząc w tym momencie jako admin nie widzę guzika do wysłania wiadomości do usera (nie wiem czemu, przecież korzystając z uprawnień administracyjnych powinienem móc obejść każdą blokadę), po wejściu na `/message/user` już mogę wysyłać czyli ok.

Podobne dziwne problemy w relacji user-inny user. Po zablokowaniu kogoś znika jednemu guzik do wysłania mu wiadomości, po wejściu ręcznie na adres też nie działa. Ta druga zablokowana osoba nadal guzik widzi, ale po kliknięciu nie może wysyłać.

Zastanawiając się nad tym jeszcze chwileczkę dostrzegam jeszcze jeden problem: zablokowanie konkretnej osoby blokuje całą listę wiadomości od niej. To chyba jest słabe, gdy ktoś kogoś zablokuje to jak najbardziej ten ktoś nie może wysyłać do niego nowych wiadomości, ale wydaje mi się, że powinien móc przejrzeć wcześniej wysłane wiadomości, w końcu do tego momentu obydwoje godzili się na tą korespondencję.

Nadal brakuje mi też listy zablokowanych, brak tego to spore utrudnienie bo założę się ze po miesiącu przeciętny użytkownik nie będzie pamiętał kogo już tam zablokował i nie wiedział kto nie może do niego napisać.

awaluk commented:
- Jestem zalogowany jako `user2`. Chcę wysłać PW do użytkownika `user3`. Nie ma żadnych blokad (ani `user2` nie zablokował `user3`, ani odwrotnie). Po wejściu na profili `user` widzę przycisk "wyślij wiadomość prywatną", niestety po kliknięciu przechodzi na stronę wysyłania wiadomości, która jest pusta
![Zrzut ekranu z 2020-03-03 21-06-54](https://user-images.githubusercontent.com/13801608/75815004-f974a100-5d92-11ea-8510-2e97f4e3fa7e.png)
- Gdy jako `user2` zablokuję PW od `user3` po wejściu na jego profil już nie widzę przycisku, a to źle bo w ten sposób nie będę w stanie zobaczyć historii rozmowy do momentu zablokowania. Teoretycznie powinno się dać w skrzynce odbiorczej, ale to raczej mało wygodne, bo tam jest lista wszystkich pomieszana. Dlatego pomyślałbym czy np. na profilu nie zmienić przycisku "wyślij wiadomość prywatną" (oraz tytułu podstron itd.) na coś bardziej ogólnego, np. "wiadomości prywatne" co zawsze kierowałoby do widoku rozmowy, gdzie ktoś mógłby zobaczyć historię/informację o braku wiadomości oraz wysłać nową wiadomość/dostać komunikat, że do tej osoby PW nie może wysłać.
- Jak już ktoś jest zablokowany, to jak dla mnie nie powinien widzieć formularza wysyłania na podstronie oraz przycisku, tym bardziej że to jakoś wyraźnie nie sygnalizuje, że jest zablokowane. Ukryłbym pole i przycisk wysyłania, bardziej uwidocznił napis, że nie można wysłać PW do tej osoby oraz nie precyzował powodu.
- Pozostawiam użytkownika zablokowanego, przelogowuję się na `user3` i wchodzę na profil `user2`. Tam jest jakiś dziwny przycisk z jeszcze bardziej dziwnym labelem i znów zablokowany, ale nie idzie tego nijak poznać. W takiej sytuacji przycisk powinien po prostu zniknąć. Z tym że tutaj blokada użytkowników powinna być od siebie niezależna - w tej chwili nie mogę zablokować tego użytkownika bo on mnie zablokował. A jak za 5 minut on mnie odblokuje to muszę tam wrócić i go dopiero zablokować? No nie, blokada z każdej strony musi działać niezależnie.
Poza tym tak jak rozmawialiśmy, nie ujawniamy czy ktoś kogoś zablokował, wyłączył PW czy dlaczego nie można wysłać PW - po prostu nie może
    > myślę ze pokazać tylko tych, których zablokowałeś. Kto Cię zablokował raczej nie powinieneś wiedzieć, bez sensu to ujawniać. I tak user zobaczy że nie może wysłać mu wiadomości, ale wtedy nie będzie wiedział czy wyłączył pw czy jest zablokowany przez tę osobę

    tak samo taki użytkownik traci dostęp do historii rozmowy, bo przycisk zniknął, bo ktoś go zablokował (on tego przypuszczalnie nie chciał)
- Loguję się jako `user2`, blokuję admina `admin`. Loguję się jako `admin`, mogę wysłać użytkownikowi PW pomimo blokady i to się zgadza. Jednak po najechaniu na przycisk wysyłania na profilu widzę informację jakoby użytkownik miał wyłączone PW od innych użytkowników, co jest nieprawdą, bo ma włączone (po odblokowaniu już komunikatu nie ma). 
- Zastanawiam się w ogóle jak rozegrać tę blokadę dla administracji... Wyżej już była o tym mowa jak widzę, ale jak teraz to testuję to lekko nie ma to sensu. Administracja może wysłać PW, użytkownik je dostanie i nie może odpowiedzieć bo sam kogoś zablokował. Skoro już może dostać od administracji najczęściej jakąś np. prośbę o wyjaśnienie to powinien móc na to odpowiedzieć. Wydaje mi się, że najlepszym wyjściem byłby brak możliwości blokady osób z administracji, może wcale nie być guzika do blokady po prostu.
- Gdy próbuję się dostać na stronę z listą zablokowanych użytkowników dostaję błąd, nie mogę więc niczego tam sprawdzić
    > Parse error: syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /vagrant/forum/qa-include/pages/pw-blocked-list.php on line 71

Z racji tylu uwag i nawet braku możliwości wysłania PW gdy użytkownik jest odblokowany, kodu na razie nie sprawdzam, bo nie ma to sensu.

Mariusz08 commented:
> Jestem zalogowany jako user2. Chcę wysłać PW do użytkownika user3. Nie ma żadnych blokad (ani user2 nie zablokował user3, ani odwrotnie). Po wejściu na profili user widzę przycisk "wyślij wiadomość prywatną", niestety po kliknięciu przechodzi na stronę wysyłania wiadomości, która jest pusta

Fixed

> Gdy jako user2 zablokuję PW od user3 po wejściu na jego profil już nie widzę przycisku, a to źle bo w ten sposób nie będę w stanie zobaczyć historii rozmowy do momentu zablokowania. Teoretycznie powinno się dać w skrzynce odbiorczej, ale to raczej mało wygodne, bo tam jest lista wszystkich pomieszana. Dlatego pomyślałbym czy np. na profilu nie zmienić przycisku "wyślij wiadomość prywatną" (oraz tytułu podstron itd.) na coś bardziej ogólnego, np. "wiadomości prywatne" co zawsze kierowałoby do widoku rozmowy, gdzie ktoś mógłby zobaczyć historię/informację o braku wiadomości oraz wysłać nową wiadomość/dostać komunikat, że do tej osoby PW nie może wysłać.

Fixed

> Jak już ktoś jest zablokowany, to jak dla mnie nie powinien widzieć formularza wysyłania na podstronie oraz przycisku, tym bardziej że to jakoś wyraźnie nie sygnalizuje, że jest zablokowane. Ukryłbym pole i przycisk wysyłania, bardziej uwidocznił napis, że nie można wysłać PW do tej osoby oraz nie precyzował powodu.

Fixed

> Pozostawiam użytkownika zablokowanego, przelogowuję się na user3 i wchodzę na profil user2. Tam jest jakiś dziwny przycisk z jeszcze bardziej dziwnym labelem i znów zablokowany, ale nie idzie tego nijak poznać. W takiej sytuacji przycisk powinien po prostu zniknąć. Z tym że tutaj blokada użytkowników powinna być od siebie niezależna - w tej chwili nie mogę zablokować tego użytkownika bo on mnie zablokował. A jak za 5 minut on mnie odblokuje to muszę tam wrócić i go dopiero zablokować? No nie, blokada z każdej strony musi działać niezależnie.
Poza tym tak jak rozmawialiśmy, nie ujawniamy czy ktoś kogoś zablokował, wyłączył PW czy dlaczego nie można wysłać PW - po prostu nie może

Fixed

> Loguję się jako user2, blokuję admina admin. Loguję się jako admin, mogę wysłać użytkownikowi PW pomimo blokady i to się zgadza. Jednak po najechaniu na przycisk wysyłania na profilu widzę informację jakoby użytkownik miał wyłączone PW od innych użytkowników, co jest nieprawdą, bo ma włączone (po odblokowaniu już komunikatu nie ma).

Fixed

> Gdy próbuję się dostać na stronę z listą zablokowanych użytkowników dostaję błąd, nie mogę więc niczego tam sprawdzić

Fixed

> Zastanawiam się w ogóle jak rozegrać tę blokadę dla administracji... Wyżej już była o tym mowa jak widzę, ale jak teraz to testuję to lekko nie ma to sensu. Administracja może wysłać PW, użytkownik je dostanie i nie może odpowiedzieć bo sam kogoś zablokował. Skoro już może dostać od administracji najczęściej jakąś np. prośbę o wyjaśnienie to powinien móc na to odpowiedzieć. Wydaje mi się, że najlepszym wyjściem byłby brak możliwości blokady osób z administracji, może wcale nie być guzika do blokady po prostu.

Na razie wyłączyłem możliwość blokowania osób z administracji.

awaluk commented:
- Wchodzę na swój profil (aktualnie zalogowanego usera) i mam przycisk "Zobacz historię prywatnych wiadomości". Po kliknięciu brak uprawnień oczywiście.
- > Nie możesz wysłać wiadomości prywatnej do tego uzytkownika.

  u**ż**ytkownika - drobna literówka, ale rzuca się w oczy

- Przypadkiem odkryłem, że ta blokada wcale nie działa. Mam odblokowane PW dla userów, normalnie. Wchodzę na stronę wiadomości prywatnych, wpisuję treść, ale jeszcze nie wysyłam. Użytkownik do którego chcę wysłać w tym momencie mnie blokuje. Wracam do formularza, klikam wyślij i PW normalnie jest wysyłane i dochodzi. Nie zagłębiałem się w kod bo sporo tam różnych warunków, ale wygląda na to, że zrobiłeś tylko ukrywanie formularza, a samo PW przy wykonaniu odpowiedniego requesta nadal można sobie bez problemu wysyłać.

Reszta problemów wydaje się rozwiązana

event15 commented on forum/qa-include/pages/message.php:
```diff
@@ -1,257 +1,288 @@
+//    Check the user exists and work out what can and can't be set (if not using single sign-on)
+        
 
-        }
```
czemu taki kawał kodu został usunięty?

event15 commented on forum/qa-include/pages/message.php:
```diff
@@ -1,257 +1,288 @@
+    $allowedPrivateMessages = qa_opt('allow_private_messages');
+    $blockedPrivateMessageBool = 0 != $blockedPrivateMessages->num_rows;
+
+    if (
```
zbędny enter

event15 commented on forum/qa-include/pages/message.php:
```diff
@@ -1,257 +1,288 @@
+                        'type' => $hideForm ? 'static' : '',
+                        'label' => qa_lang_html_sub('misc/message_for_x', qa_get_one_user_html($handle, false)),
+                        'tags' => 'name="message" id="message"',
+                        'value' => qa_html(@$inmessage, $messagesent),
```
@ do wywalenia

Mariusz08 replied:
To jest kod q2a, niezbyt chce go ruszać.

event15 replied:
do jest dodane przez Ciebie więc nie q2a

event15 commented on forum/qa-include/pages/message.php:
```diff
@@ -1,257 +1,288 @@
+                        'value' => qa_html(@$inmessage, $messagesent),
+                        'rows' => 8,
+                        'note' => qa_lang_html_sub('misc/message_explanation', qa_html(qa_opt('site_title'))),
+                        'error' => qa_html(@$errors['message']),
```
@ do wywalenia

Mariusz08 replied:
To jest kod q2a, niezbyt chce go ruszać.

event15 commented on forum/qa-include/pages/message.php:
```diff
@@ -1,257 +1,288 @@
+                        'value' => qa_html(@$inmessage, $messagesent),
+                        'rows' => 8,
+                        'note' => qa_lang_html_sub('misc/message_explanation', qa_html(qa_opt('site_title'))),
+                        'error' => qa_html(@$errors['message']),
```
@ do wywalenia

Mariusz08 replied:
To jest kod q2a, niezbyt chce go ruszać.

event15 commented on forum/qa-include/pages/pw-blocked-list.php:
```diff
@@ -0,0 +1,94 @@
+    
+    $userId = qa_get_logged_in_userid();
+    if (empty($userId)) {
+        $qa_content['error'] = 'Musisz być zalogowany aby wykonać tą czynność';
```
brak przecinka przed aby

event15 commented on forum/qa-include/pages/pw-blocked-list.php:
```diff
@@ -0,0 +1,94 @@
+    
+    $blockedUsers = qa_db_select_with_pending([
+                'columns' => ['^users.userid', '^users.handle',  '^users.flags', '^users.email', 'avatarblobid' => 'BINARY avatarblobid', '^users.avatarwidth', '^users.avatarheight'],
+                'source' => '^users JOIN (SELECT toUserId FROM ^blockedpw WHERE fromUserId = #) s ON ^users.userid=s.toUserId',//'^users JOIN (SELECT userid FROM ^userpoints ORDER BY points DESC LIMIT #,#) y ON ^users.userid=y.userid JOIN ^userpoints ON ^users.userid=^userpoints.userid',
```
niepotrzebny zakomentowany kod

event15 commented on forum/qa-include/pages/pw-blocked-list.php:
```diff
@@ -0,0 +1,94 @@
+    if (0 === count($blockedUsers)) {
+        $pageContent = 'Nikogo jeszcze nie zablokowałeś, ale gdy zajdzie taka potrzeba, nie wahaj się';
+    } else {
+    
```
niepotrzebny enter

event15 commented on forum/qa-include/pages/pw-blocked-list.php:
```diff
@@ -0,0 +1,94 @@
+    $qa_content['navigation']['sub'] = qa_user_sub_navigation(qa_get_logged_in_handle(), 'blocklist', true);
+    
+    
+    return $qa_content;
```
brak nowej linii na końcu pliku

event15 commented on forum/qa-include/pages/user-profile.php:
```diff
@@ -270,7 +270,11 @@
     }
 
-
+    if (qa_clicked('doblockpw')) {
```
ogolnie to nie jestem pewien, ale wydaje mi się, że nie powinno się nadpisywać qa-include/pages/* skoro mechanizm tworzenia pluginów takie coś umożliwia: https://docs.question2answer.org/plugins/layers/

event15 replied:
https://docs.question2answer.org/plugins/modules-page/

Mariusz08 replied:
Rozważaliśmy wprowadzenie tego ficzeru jako wtyczka, ale mechanizm pluginów nie oferował jakiejś istotnej funkcji do rozwiązania tego problemu, więc mieliśmy dwa wyjścia:
-albo zrobić połowę kodu w pluginie a połowę w silniku bezpośrednio
-albo wszystko robić w silniku
Wybrałem to drugie, nie wyobrażam sobie żeby część była w pluginie a część nie.

event15 replied:
nie widzę w historii rozmowy tych rozważań - nie wiem co to za funkcja, ale jestem niemal pewien, że da się to zrobić bez ingerencji w core

Mariusz08 replied:
No tutaj tego nie ma, rozmawialiśmy chyba gdzieś na IRC jeszcze o tym, generalnie większość tego kodu da się przenieść do wtyczki, problem pojawia się z tego co pamiętam przy przechwytywaniu wysyłania wiadomości oraz renderowaniu custom boxa w `message.php`

event15 replied:
@awaluk odniesiesz się? 
Moim zdaniem każdy kawałek kodu można zrobić bez ingerencji w core q2a. 
Change my mind

awaluk replied:
Ciężko mi się odnieść. Czy była taka rozmowa i takie próby to nie pamiętam, zbyt długo to nie było ruszane. Czy każdy kawałek kodu można nadpisać też nie wiem - nie zaprzeczę i nie potwierdzę. Myślę że większość się da, ale czy na pewno _każdy_ to się nie odważę powiedzieć. Być może też nawet jeśli się da to może być to znacznie bardziej trudne i czasochłonne niż zmienić bezpośrednio. Jeśli powiedzmy to zwiększa czas pracy nad zmianą dwukrotnie to ok, ale jak zwiększa pięciokrotnie to już nie wiem czy na pewno ma to sens.

W związku z tym zgadzam się jak najbardziej, że najlepiej, aby zmiany były na pluginach, fajnie jakbyśmy tak próbowali, ale nie mogę ze 100% pewnością powiedzieć, że to się zawsze uda i będzie najlepszym rozwiązaniem. Zgodzić się też wypada z Mariuszem, że jeśli plugin ma być niekompletny to niech już lepiej wcale nie będzie pluginem, bo ktoś go kiedyś go wyciągnie z zamysłem że to normalny plugin i się posypie. Jeśli uda się całość przenieść na plugin to jak najbardziej fajnie.

event15 replied:
> Jeśli powiedzmy to zwiększa czas pracy nad zmianą dwukrotnie to ok, ale jak zwiększa pięciokrotnie to już nie wiem czy na pewno ma to sens.
Oczywiście, że ma sens. Nie ma tu jiry, żeby sobie trackować czas. Jest pełno PR, które wiszą tu kilka lat i nikt nie ma z tym problemu.

awaluk replied:
Jak dla mnie nie ma. Nikt nie ma problemu, bo nikt nie wymaga od nikogo pracy nad daną rzeczą. Jak ktoś coś robił, kolokwialnie mówiąc mu się odechciało i nie robi no to ok, rozumiem to, nie płacimy za pracę, nie ma obowiązku. Ale jeśli już ktoś chce coś dla nas robić i zapyta mnie, to niech robi coś co ma sens.

Przykładowo: jak ktoś ma spędzić nad zrobieniem 1 rzeczy jako plugin tydzień, albo w tym czasie może zrobić tę 1 rzecz bez plugina w dzień i do tego zrobić jeszcze 5 innych rzeczy w pozostałe dni tygodnia, to wybrałbym wariant pierwszy. Oczywiście zależy od sytuacji, złożoności problemu, konkretnych potrzeb. Super byłoby aby ten ktoś zrobił plugin, ale jak w tym czasie nadal zrobi to dobrze tylko że bez pluginu i dostarczy nam 5 innych potrzebnych rzeczy no to sorry, ale jednak zysk funkcjonalny jest znacznie większy i zyskają użytkownicy na co dzień korzystający z forum. Też bym chciał żeby wszystko było idealnie, ale nie zawsze to jest najlepsze wyjście gdy ma sie ograniczone zasoby i czas.

event15 replied:
> Przykładowo: jak ktoś ma spędzić nad zrobieniem 1 rzeczy jako plugin tydzień, albo w tym czasie może zrobić tę 1 rzecz bez plugina w dzień i do tego zrobić jeszcze 5 innych rzeczy w pozostałe dni tygodnia, to wybrałbym wariant pierwszy.

Aż nie mogę wyjść z podziwu, że takie masz zdanie mając na uwadze fakt, że od ponad 30 godzin walczę nad jednym plikiem który był zrobiony na kolanie BEZ PLUGINU bo było szybciej. 
Nie po to twórca q2a stworzył możliwość tworzenia praktycznie wszystkiego niezależnie bez ingerencji w kod projektu żeby teraz wrzucać tam co się chce. Przez takie akcje aktualizacja wersji forum - a co za tym idzie, przyspieszenie jego działania, zwiększenie bezpieczeństwa i nowe funkcjonalności które są dostępne bez pluginów - staje się na tyle uciążliwa, że niemal niemożliwa. 

Serio wygoda i szybkość wdrożenia zmiany mają być jedynymi argumentami w tej dyskusji? Nie ma przymusu - skoro ktoś zaczął coś robić, to niech chociaż robi tak, żeby nie utrudniać innym pracy. Po co mi milion funkcjonalności w kodzie, które są naszyte w całym core, skoro później muszę zapewne miesiąc jak i nie dłużej walczyć z upgrade wersji forum? Skoro są to dodatkowe funkcjonalności to nie są "must have" a skoro nie są must have, to niech chociaż będą przyjazne do wdrożenia i rozwijania - nawet kosztem czasu. Masa rzeczy tu wisi latami i dalej nie jestem przekonany do Twojej argumentacji. 
Sam wiesz, że nic tu się nie zyskuje słabymi rozwiązaniami, a otrzymuje się w zamian straszny bagaż.

awaluk replied:
Nigdzie nie napisałem, że szybkość i wygoda to jedyne argumenty. Przeciwnie - kilkukrotnie zaznaczyłem, że najlepiej by było, aby było to zrobione jako plugin i należałoby do tego dążyć. Jednak również ważny jest czas i możliwość wprowadzenia zmian, a Ty zdaje się, że jako jedyny argument widzisz kod. Trzeba to wypośrodkować - robić możliwie najlepszy kod, jednocześnie w sensownym czasie i z rozsądnym użyciem "zasobów ludzkich" jakie mamy (a przez kilka ostatnich miesięcy nie było w zasadzie żadnych).

event15 replied:
Nie mówię o możliwie najlepszym kodzie - bo to jest kwestia zupełnie niezwiązana z moim postulatem. 
Chodzi o wydzielenie kodu tak by całość znajdowała się w pluginie. 
Żeby tak się stało, trzeba przeczytać dokumentację samego q2a, w której wszystko jest opisane:
https://docs.question2answer.org/plugins/modules/
https://docs.question2answer.org/plugins/modules-page/

Wystarczy pogrzebać. Nie jest to najbardziej dopracowana pod kątem UX dokumentacja, ale każdą zmianę można w prosty i szybki sposób zrobić tak, by nie ingerować w kod core. 
Nie ma problemu z tworzeniem nowych pages i przechwytywaniem z nich requestów, 
Nie ma problemu z przechwytywaniem requestów z istniejących pages. 
Można nawet sugerować strony. 

Można też nadpisać style bez używania qa-styles.css domyślnego motywu. 
W zasadzie dokumentacja pokazuje dokładnie w jaki sposób zrobić cokolwiek sie tylko chce - jest nawet 150 metod które można własnoręcznie nadpisać z core przy pomocy pluginu. 

Nie widzę żadnej potrzeby, konieczności ani trudności w tworzeniu zmian, które by nie wpływały na core. Jakość kodu w tym przypadku jest drugorzędna i jest rozwiązywana w trakcie pracy nad PR. 

Problemem nie jest samo tworzenie zmiany tylko nieczytanie dokładnie dokumentacji - co by masę rzeczy uprościło i przyspieszyło. Nie trzeba by robić haków na core, które zwiększają ryzyko wysypania się forum, którego nie da się przywrócić tylko wyłączeniem pluginu. Gdy zmieniamy coś w core to nie podziała wyłączenie pluginu tylko rollback zmian na kodzie który nie powinien być zmieniany - a niestety ma mase zmian, których nie powinien mieć. Szkoda pogłębiać tę przepaść. 

Idąc tym tokiem rozumowania - miesiąc poświęcony na aktualizację wersji może być czasem poświęconym bez sensu, bo wolałbyś, żebym zrobił 10 issues z tego projektu nawet jeśli miałyby być zrobione w core?

awaluk replied:
Przeinaczasz na siłę to co chcę przekazać :) Zupełnie niepotrzebnie, bo ja naprawdę chciałbym jak najlepiej i poprawić to jak się da, a najlepiej cytując klasyka "zaorać" i zrobić od nowa. Tylko że się tak łatwo nie da. Chodzi mi o to, abyśmy nie utknęli na przenoszeniu prawie gotowych zmian na pluginy. Nie wiem na ile to jest trudne, bo nie zagłębiałem się w ten kod, do tego nie znam wszystkich możliwych przypadków i Ty też nie znasz. Dziwi mnie nieco więc, że wypowiadasz się z taką pewnością. Trochę rzeczy już w Q2A robiłem i nie było żadnego problemu z pluginami, z pewnymi miałem różnego rodzaju problemy. Być może problemy były z mojej winy, nie wiem, być może źle szukałem, co nie zmienia faktu że mogą się pojawić i może to potrwać, tym bardziej jak ktoś nie ma wprawy w dłubaniu w Q2A.

W związku z tym jeśli teraz znowu na miesiące zawiśniemy z prawie gotowymi zmianami, a może w ogóle nigdy nie wejdą jak komuś nie będzie chciało się zrobić/nie będzie wiedział jak, to tracimy coś, co moglibyśmy mieć. I oczywiście, najlepiej byłoby to zrobić jako plugin, tak, będzie super jak tak zrobimy. Ale nie ukrywajmy, że nie ma chętnych do pracy przy tym kodzie. Sam średnio miałem na to czas, chęci i możliwości, tak jak i Ty przez ostatnie miesiące nie interesowałeś się zupełnie tym co się tu działo. Cieszę się, że wróciłeś i coś się zaczyna dziać, wszyscy jakoś się zebrali i idzie, ale nie wiemy jak długo. W związku z tym wolę podchodzić tak, aby robić to na pluginach, ale jak się nie uda to próbować bez plugina. Próbować możliwie najlepiej, ale gdy się nie da to spróbować nieco gorzej, żeby mieć cokolwiek niż nie mieć nic. Bo nie ma kto tego niestety robić, nie mamy zatrudnionych ludzi.

> Nie widzę żadnej potrzeby, konieczności ani trudności w tworzeniu zmian, które by nie wpływały na core.

Cieszę się, że Ty nie widzisz, ale inni mogą mieć pewne trudności, sam nie raz je miałem.

> Idąc tym tokiem rozumowania - miesiąc poświęcony na aktualizację wersji może być czasem poświęconym bez sensu, bo wolałbyś, żebym zrobił 10 issues z tego projektu nawet jeśli miałyby być zrobione w core?

Zacytuję samego siebie:
> **Trzeba to wypośrodkować** - robić możliwie najlepszy kod, jednocześnie w sensownym czasie i z rozsądnym użyciem "zasobów ludzkich" jakie mamy

Aktualizacja silnika forum przynosi nowe funkcje, zmiany, optymalizacje, poprawki błędów - więc nie jest na pewno czasem poświęconym bez sensu. 10 issues wprowadziłoby pewnie wiele nowości i potrzebnych zmian, tak jak i aktualizacji forum wprowadzi wiele nowości i potrzebnych zmian.

To co podawałem to jak mówiłem przykłady, w tym konkretnym pull requeście nie wiem ile trzeba na przeniesienie tego na plugin, może w 15 minut autor to zrobi i cała gadka jest zbędna. Proponuję więc skończyć dyskusję, bo nic ona już chyba nie wniesie. Nie przekonasz mnie, że jedyną wartością ma być sposób/jakość pisanego kodu, a czas i koszt wykonania jest nieważny - to zawsze w jakimś stopniu jest ważne, trzeba znaleźć dobry środek. Moja propozycja: starajmy się wszystko robić na pluginach, ale jeśli się nie komuś nie uda, a dana funkcjonalność jest fajna/potrzebna i nie ma innego pomysłu jak to zrobić bez pluginu oraz innej osoby, która mogłaby to przejąć i zrobić jako plugin, miejmy z tyłu głowy pomyślenie o opcji dodania czegoś bez plugina. Ale tylko gdzieś z tyłu, domyślnie myślmy o wszystkim jako pluginie. Tylko tyle, proste.

event15 commented on forum/qa-include/qa-index.php:
```diff
@@ -165,11 +165,13 @@ function qa_index_set_request()
+			require QA_INCLUDE_DIR . 'qa-install.php';
+		} elseif ($requestlower == 'url/test/'.QA_URL_TEST_STRING) {
+			require QA_INCLUDE_DIR . 'qa-url-test.php';
+		} elseif ($requestlower == 'blocked-users-list') {
```
poszukaj sposobu na nadpisanie tego bez nadpisywania core q2a. Wolałbym to zmienić w pluginie.

Mariusz08 replied:
To akurat da się zrobić w plugnie (w sensie custom page) ale tak jak wspominałem wyżej nie ma to tutaj sensu.

event15 replied:
jeżeli da się to zrobić w pluginie, to bezapelacyjnie musi być w pluginie

event15 commented on forum/qa-include/qa-page-users.php:
```diff
@@ -2,5 +2,5 @@
  * @deprecated This file is deprecated from Q2A 1.7; use the below file instead.
  */
-
+ 
```
cofnij te zmiany

event15 commented on forum/qa-include/qa-page.php:
```diff
@@ -473,6 +473,7 @@ function qa_page_routing()
 			'users' => 'pages/users.php',
 			'users/blocked' => 'pages/users-blocked.php',
 			'users/special' => 'pages/users-special.php',
+			'blocked-users-list' => 'pages/pw-blocked-list.php'
```
moim zdaniem powinien to robić plugin, zmiany w core q2a to ostateczność jeśli już nie da się uniknąć takich zmian.

Mariusz08 replied:
jw

event15 commented on forum/qa-include/qa-pw-blocked-list.php:
```diff
@@ -0,0 +1,3 @@
+<?php
+
+require_once QA_INCLUDE_DIR . 'pages/pw-blocked-list.php';
```
brak nowej linii oraz to powinno być częścią pluginu

Mariusz08 replied:
jw

Mariusz08 commented:
Robimy z tego plugin, żeby nie mieszać zbytnio tworzę nowy PR. Ten do utylizacji.
